pennylane_lightning/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp

```diff
@@ -0,0 +1,766 @@
+    }
+
+    /**
+     * @brief Calculate expectation value with respect to two qubit observable
+     * on specified wires.
+     *
+     * @param matrix Hermitian matrix representing observable to be used.
+     * @param wires Wires to apply observable to.
+     * @param params Not used.
+     * @return Expectation value with respect to observable applied to specified
+     * wires.
+     */
+    auto getExpectationValueTwoQubitOp(
+        const KokkosVector &matrix, const std::vector<size_t> &wires,
+        [[maybe_unused]] const std::vector<PrecisionT> &params = {0.0}) {
+        const size_t num_qubits = this->_statevector.getNumQubits();
+        Kokkos::View<ComplexT *> arr_data = this->_statevector.getView();
+        PrecisionT expval = 0;
+        Kokkos::parallel_reduce(
+            exp2(num_qubits - 2),
+            getExpectationValueTwoQubitOpFunctor<PrecisionT>(
+                arr_data, num_qubits, matrix, wires),
+            expval);
+        return expval;
+    }
+
+    /**
+     * @brief Calculate expectation value with respect to multi qubit observable
+     * on specified wires.
+     *
+     * @param matrix Hermitian matrix representing observable to be used.
+     * @param wires Wires to apply observable to.
+     * @param params Not used.
+     * @return Expectation value with respect to observable applied to specified
+     * wires.
+     */
+    auto getExpectationValueMultiQubitOp(
+        const KokkosVector &matrix, const std::vector<size_t> &wires,
+        [[maybe_unused]] const std::vector<PrecisionT> &params = {0.0}) {
+        if (wires.size() == 1) {
+            return getExpectationValueSingleQubitOp(matrix, wires, params);
+        } else if (wires.size() == 2) {
+            return getExpectationValueTwoQubitOp(matrix, wires, params);
+        } else {
+            Kokkos::View<const size_t *, Kokkos::HostSpace,
+                         Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+                wires_host(wires.data(), wires.size());
+            const size_t num_qubits = this->_statevector.getNumQubits();
+            const Kokkos::View<ComplexT *> arr_data =
+                this->_statevector.getView();
+
+            Kokkos::View<size_t *> wires_view("wires_view", wires.size());
+            Kokkos::deep_copy(wires_view, wires_host);
+            PrecisionT expval = 0;
+            Kokkos::parallel_reduce(
+                Kokkos::RangePolicy<KokkosExecSpace>(
+                    0, exp2(num_qubits - wires.size())),
+                getExpectationValueMultiQubitOpFunctor(arr_data, num_qubits,
+                                                       matrix, wires_view),
+                expval);
+            return expval;
+        }
+    }
+
+    /**
+     * @brief Calculate expectation value for a general Observable.
+     *
+     * @param ob Observable.
+     * @return Expectation value with respect to the given observable.
+     */
+    auto expval(const Observable<StateVectorT> &ob) {
```

Comment:
As far as I know, LQubit doesn't have special `expval` functions like `getExpectationValuePauliX`. With Kokkos, it made sense to do a few like that because on GPUs you can reach about twice the bandwidth with a parallel reduce compared with a parallel for over a vector. You thus want to apply the gate and reduce on the fly. In LQubit, we make a state vector copy, apply the gate and then perform an inner product. I think it could also benefit from reducing on the fly, but it should be tested. 

I should also note that LKokkos expval kernels look like 
```
        expval += real(
            conj(arr[i0]) * (matrix[0B00] * arr[i0] + matrix[0B01] * arr[i1]) +
            conj(arr[i1]) * (matrix[0B10] * arr[i0] + matrix[0B11] * arr[i1]));
```
We could reduce the number of operations by taking `real` analytically instead of performing the whole thing with complex numbers and finally taking the real part. We could also take advantage of the fact that `matrix` is Hermitian I think.

Reply:
(sorry) More to your point, LKokkos didn't have any `expval` before and the bindings were pointing to `getExpectationValue...` which is why I wanted to keep them for now. Once bindings are properly integrated and tested, I would simply merge/remove all `getExpectationValue...` into `expval` overloads.

Reply:
Awesome, sounds good. Thanks @vincentmr